src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/StorePal.AppleKeychainStore.cs

```diff
@@ -68,13 +68,32 @@ public void Add(ICertificatePal cert)
 
             public void Remove(ICertificatePal cert)
             {
-                if (_readonly)
-                    throw new CryptographicException(SR.Cryptography_X509_StoreReadOnly);
-
+                // First check if the certificate is even in the keychain
                 AppleCertificatePal applePal = (AppleCertificatePal)cert;
-
-                var handle = (SafeKeychainItemHandle)applePal.IdentityHandle ?? applePal.CertificateHandle;
-                Interop.AppleCrypto.X509StoreRemoveCertificate(handle, _keychainHandle);
+                using (SafeCFArrayHandle certs = Interop.AppleCrypto.KeychainEnumerateCerts(_keychainHandle))
+                {
+                    long count = Interop.CoreFoundation.CFArrayGetCount(certs);
+
+                    for (int i = 0; i < count; i++)
+                    {
+                        IntPtr itemHandle = Interop.CoreFoundation.CFArrayGetValueAtIndex(certs, i);
+
+                        SafeSecCertificateHandle certHandle;
+                        SafeSecIdentityHandle identityHandle;
+
+                        if (Interop.AppleCrypto.X509DemuxAndRetainHandle(itemHandle, out certHandle, out identityHandle))
+                        {
+                            if (applePal.IdentityHandle == identityHandle && applePal.CertificateHandle == certHandle)
```

Comment:
Is pointer identity guaranteed here?

Reply:
Good question, I'm not entirely sure. I was kinda hoping that the outer loop tests would rule that out, since I was expecting them to fail. I'll see if I can write a test where I first add a certificate to the store and then try to remove it and then check if it is still in the store. Or perhaps just checking the thumbprint of the certificate would do the trick, but I think that might cause a lot of overhead?